### PR TITLE
Feature/modals [#171407712]

### PR DESCRIPTION
**Pull request by lus21**
[Story link](https://www.pivotaltracker.com/story/show/171407712)

- Create 3 Modal UI
   - Pass Stipulation
    - Incomplete Stipulation
     - Waive Stipulation
- connected with redux

- hitting on CTA buttons open corresponding modal

### CODE REVIEW

File: components/Dashboard/VerificationsContainer/CtaFooter.tsx [hunk 1]
```diff
@@ -4,26 +4,57 @@ import { Box } from '@material-ui/core';
...
-        <span className={classes.icon}>
-          <Info />
-        </span>
+
+type Props = IWithStylesProps & DispatchProps;
+
+const ctaFooterWithStyles = withStyles(CtaFooterStyles)((props: Props) => {
+  const { classes, toggleModal } = props;
+
+  const handleButtonAction = (modalType: ModalType) => {
```
Comment (eryash12) [inline components/Dashboard/VerificationsContainer/CtaFooter.tsx]:
```suggestion
  const handleButtonAction = (modalType: ModalType) => () => {
```

File: components/Dashboard/VerificationsContainer/CtaFooter.tsx [hunk 2]
```diff
@@ -4,26 +4,57 @@ import { Box } from '@material-ui/core';
...
+    toggleModal({ modalType });
+  };
+
+  return (
+    <Paper square elevation={4} className={classes.container}>
+      <Box display="flex" justifyContent="space-between">
+        <Box className={classes.buttons}>
+          <Button color={passButton} onClick={() => handleButtonAction(ModalTypes.PassStipulation)}>
```
Comment (eryash12) [inline components/Dashboard/VerificationsContainer/CtaFooter.tsx]:
```suggestion
          <Button color={passButton} onClick={handleButtonAction(ModalTypes.PassStipulation)}>
```

File: components/Dashboard/VerificationsContainer/CtaFooter.tsx [hunk 3]
```diff
@@ -4,26 +4,57 @@ import { Box } from '@material-ui/core';
...
+            Pass
+          </Button>
+          <Button color={incompleteButton} onClick={() => handleButtonAction(ModalTypes.IncomepleteStipulation)}>
```
Comment (eryash12) [inline components/Dashboard/VerificationsContainer/CtaFooter.tsx]:
```suggestion
          <Button color={incompleteButton} onClick={handleButtonAction(ModalTypes.IncomepleteStipulation)}>
```

File: components/Dashboard/VerificationsContainer/CtaFooter.tsx [hunk 4]
```diff
@@ -4,26 +4,57 @@ import { Box } from '@material-ui/core';
...
+            Incomplete
+          </Button>
+          <Button color={waiveButton} onClick={() => handleButtonAction(ModalTypes.WaiveIncompleteStipilation)}>
```
Comment (eryash12) [inline components/Dashboard/VerificationsContainer/CtaFooter.tsx]:
```suggestion
          <Button color={waiveButton} onClick={handleButtonAction(ModalTypes.WaiveIncompleteStipilation)}>
```

File: actions/createors.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+export function createActionFunction<TAction extends Action<TAction['type']>>(
+  type: TAction['type'],
+): () => Action<TAction['type']> {
+  return () => ({
+    type,
+  });
+}
+
+export const createSetModalAction = createPayloadedActionFunction<ActionSetModal>(SET_MODAL);
```
Comment (eryash12) [inline actions/createors.ts]: Can you teach me Generics please?? :P

File: components/Dashboard/VerificationsContainer/CtaFooter.tsx [hunk 6]
```diff
@@ -4,26 +4,57 @@ import { Box } from '@material-ui/core';
...
-));
+    </Paper>
+  );
+});
+
+interface DispatchProps {
+  toggleModal: (payload: SetModalPayload) => void;
+}
+
+function mapDispatchToProps(dispatch: Dispatch<any>): DispatchProps {
```
Comment (eryash12) [inline components/Dashboard/VerificationsContainer/CtaFooter.tsx]: Does the Dispatch type imported from react match with React-redux?

File: components/Modal/IncompleteFooter.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
+import Box from '@material-ui/core/Box';
+import Button from 'components/CustomMui/Button';
+import { createStyles, withStyles } from '@material-ui/core/styles';
+import TextField from '@material-ui/core/TextField';
+import React from 'react';
+import { colors } from '../../styles/main';
+import { IWithStylesProps } from '../../types/main';
+import { modalFontStyles } from './modalStyles';
+
+const incompleteFooterStyles = () =>
```
Comment (eryash12) [inline components/Modal/IncompleteFooter.tsx]: Why not create a ModalStyles file and dump all the styles there, will be cleaner right?

Comment (eryash12) [reply components/Modal/IncompleteFooter.tsx]: Sorry you already have that, just throw this there

File: components/Modal/IncompleteStipulation.tsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+import { createStyles, withStyles } from '@material-ui/core/styles';
+import Typography from '@material-ui/core/Typography';
+import CheckBoxOutlinedIcon from '@material-ui/icons/CheckBoxOutlined';
+import { Component } from 'react';
+import missingDocuments from '../../dummyData/missingDocuments';
+import { IWithStylesProps } from '../../types/main';
+import IncompleteFooter from './IncompleteFooter';
+import { modalFontStyles } from './modalStyles';
+
+const incompleteStipulationStyles = () =>
```
Comment (eryash12) [inline components/Modal/IncompleteStipulation.tsx]: same as ^^^

File: components/Modal/ModalHeader.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import { createStyles, withStyles } from '@material-ui/core/styles';
+import { IWithStylesProps } from '../../types/main';
+import { modalFontStyles } from './modalStyles';
+
+interface ParentProps {
+  title: string;
+  subtitle?: string;
+}
+
+const modalHeaderStyles = () =>
```
Comment (eryash12) [inline components/Modal/ModalHeader.tsx]: drop into modalStyles

File: components/Modal/PassStipulation.tsx [hunk 10]
```diff
@@ -0,0 +1,53 @@
+import Box from '@material-ui/core/Box';
+import Button from 'components/CustomMui/Button';
+import { createStyles, withStyles } from '@material-ui/core/styles';
+import Typography from '@material-ui/core/Typography';
+import { colors } from 'styles/main';
+import { IWithStylesProps } from '../../types/main';
+import { modalFontStyles } from './modalStyles';
+
+const passStipulationStyles = () =>
```
Comment (eryash12) [inline components/Modal/PassStipulation.tsx]: same

File: components/Modal/WaiveStipulation.tsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+import { createStyles, withStyles } from '@material-ui/core/styles';
+import Typography from '@material-ui/core/Typography';
+import Button from 'components/CustomMui/Button';
+import { colors } from 'styles/main';
+import waiveDocuments from '../../dummyData/waiveDocuments';
+import { IWithStylesProps } from '../../types/main';
+import IncompleteFooter from './IncompleteFooter';
+import { modalFontStyles } from './modalStyles';
+
+const waiveStipulationStyles = () =>
```
Comment (eryash12) [inline components/Modal/WaiveStipulation.tsx]: same

File: components/Modal/WaiveStipulation.tsx [hunk 12]
```diff
@@ -0,0 +1,84 @@
+import Box from '@material-ui/core/Box';
+import { createStyles, withStyles } from '@material-ui/core/styles';
+import Typography from '@material-ui/core/Typography';
+import Button from 'components/CustomMui/Button';
+import { colors } from 'styles/main';
+import waiveDocuments from '../../dummyData/waiveDocuments';
```
Comment (eryash12) [inline components/Modal/WaiveStipulation.tsx]: just do absolute imports like `dummyData/waiveDocuments'

File: actions/createors.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
+import { Action } from 'redux';
```
[the remark this answers is not in the document]
Comment (lus21) [reply actions/createors.ts]: this typo doesn't exist on overcoming PR as the content of this file has been moved to other files and file itself has been deleted, so let's leave it as it is here

File: components/Dashboard/VerificationsContainer/CtaFooter.tsx (hunk 6 above)
Comment (lus21) [reply components/Dashboard/VerificationsContainer/CtaFooter.tsx]: yes, I guess, React-redux has no exported member Dispatch so we should use React Dispatch as type,  it's defined like this
`type Dispatch<A> = (value: A) => void;`

the dispatch takes one argument with any type and doesn't return a value

File: actions/createors.ts (hunk 5 above)
Comment (lus21) [reply actions/createors.ts]: this mechanism is created here to have 3 types of actions
1. without any payload
2. with payload
3. with the error message

precisely what will be in the payload is depending on the action. Then in reducer in switch case, typescript will understand the corresponding payload for every type
Hope, this was helpful)

**Pull request merged**